### DISCUSSION TITLE
How does the defer work

### DISCUSSION THREAD

**Discussion by 19Qingfeng** [remix-run/remix#6634] [Q&A]
I've been working on Remix recently.

I'm curious about how Remix passes the Promise on the server to the client through defer, and Promise can't be serialized.

Does it look like a serializable object for each Promise construct in defer is passed to the client, and how should the client be notified of a state change in the Promise on the server?

I'm a little confused, and of course I know that flipping through the source code is the best answer. But I still hope that it would be better if someone can provide some relevant articles or materials. I'm much obliged

**Comment by lili21** [comment]
You can check this [gist](https://gist.github.com/jacob-ebey/9bde9546c1aafaa6bc8c242054b1be26) created by @jacob-ebey.
